x-pack/plugins/security_solution/server/endpoint/utils/action_list_helpers.ts

```diff
@@ -0,0 +1,188 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { Logger } from '@kbn/core/server';
+// eslint-disable-next-line @kbn/eslint/no-restricted-paths
+import type { SearchRequest } from '@kbn/data-plugin/public';
+import type * as estypes from '@elastic/elasticsearch/lib/api/typesWithBodyKey';
+import type { TransportResult } from '@elastic/elasticsearch';
+import { AGENT_ACTIONS_RESULTS_INDEX } from '@kbn/fleet-plugin/common';
+import { EndpointError } from '../../../common/endpoint/errors';
+import type { SecuritySolutionRequestHandlerContext } from '../../types';
+import type {
+  EndpointAction,
+  LogsEndpointAction,
+  ActionListApiResponse,
+  EndpointActionResponse,
+  LogsEndpointActionResponse,
+} from '../../../common/endpoint/types';
+import {
+  ACTIONS_SEARCH_PAGE_SIZE,
+  ACTION_REQUEST_INDICES,
+  ACTION_RESPONSE_INDICES,
+} from '../services/actions/constants';
+import { getDateFilters, logsEndpointActionsRegex } from '../services/actions/utils';
+import { catchAndWrapError } from './wrap_errors';
+import { doesLogsEndpointActionsIndexExist } from './yes_no_data_stream';
+import { GetActionDetailsListParam } from '../services/actions/action_list';
+import { ENDPOINT_ACTIONS_INDEX } from '../../../common/endpoint/constants';
+
+const queryOptions = {
+  headers: {
+    'X-elastic-product-origin': 'fleet',
+  },
+  ignore: [404],
+};
+
+export const getTimeSortedActionDetails = (
+  data: ActionListApiResponse['data']
+): ActionListApiResponse['data'] => {
+  return data.sort((a, b) => (new Date(b.startedAt) > new Date(a.startedAt) ? 1 : -1));
+};
+
+export const getActionsForAgents = async ({
+  actionTypes,
+  context,
+  elasticAgentIds,
+  endDate,
+  from,
+  logger,
+  size,
+  startDate,
+  userIds,
+}: GetActionDetailsListParam): Promise<{
+  actionIds: string[];
+  actionRequests: TransportResult<
+    estypes.SearchResponse<EndpointAction | LogsEndpointAction>,
+    unknown
+  >;
+}> => {
+  const additionalFilters = [];
+  if (actionTypes?.length) {
+    additionalFilters.push({
+      terms: {
+        'data.command': actionTypes,
+      },
+    });
+  }
+  if (userIds?.length) {
+    additionalFilters.push({ terms: { user_id: userIds } });
+  }
+  if (elasticAgentIds?.length) {
+    additionalFilters.push({ terms: { agents: elasticAgentIds } });
+  }
+
```

Comment:
should these defined `{ term: { property: string } }` for those that only defined one value? (I think I remember @pzl telling me that it would result in better ES performance)

Reply:
Not sure if that's true. I couldn't find anything in the docs that states this difference. Maybe @pzl can enlighten us. I am open to changing this to use a `term` query for a single `agentId` vs `terms` query for multiples ids if there is a significant performance improvement.

Reply:
In a traditional DB, that might be true (i.e. `x == y` instead of `x IN (y)` ). I'm not sure it's a significant difference in ES. Likely negligible at this scale

Reply:
I see. I'll leave it as it is for now.